sycl/include/CL/sycl/accessor.hpp

```diff
@@ -514,10 +519,111 @@ class image_accessor
         "Read API is not implemented on host.");
 #endif
   }
+};
+
+template <typename DataT, int Dimensions, access::mode AccessMode,
+          access::placeholder IsPlaceholder>
+class __image_slice__ {
 
-  // Available only when: accessTarget == access::target::image_array &&
-  // dimensions < 3
-  //__image_array_slice__ operator[](size_t index) const;
+  static_assert(Dimensions < 3,
+                "Image slice cannot have more then 2 dimensions");
+
+public:
+  __image_slice__(accessor<DataT, Dimensions, AccessMode,
+                           access::target::image_array, IsPlaceholder>
+                      BaseAcc,
+                  size_t Idx)
+      : MBaseAcc(BaseAcc), MIdx(Idx) {}
+
+  template <typename CoordT, int Dims = Dimensions,
+            typename = detail::enable_if_t<
+                (Dims > 0) && (IsValidCoordDataT<Dims, CoordT>::value)>>
+  DataT read(const CoordT &Coords) const {
+    constexpr int AdjustedDims = Dims == 2 ? 4 : Dims + 1;
+    using CoordElemType = typename detail::TryToGetElementType<CoordT>::type;
+
+    CoordElemType ZCoord = 0;
+
+    if(std::is_same<float, CoordElemType>::value) {
+      sycl::vec<int, Dims + 1> Size = MBaseAcc.getCountInternal();
+      ZCoord = MIdx / (float)Size.template swizzle<Dims>();
+    }
+    else {
+      ZCoord = MIdx;
+    }
+
+    sycl::vec<CoordElemType, Dims> LeftoverCoords{ZCoord};
+    sycl::vec<CoordElemType, AdjustedDims> AdjustedCoords{Coords,
+                                                          LeftoverCoords};
+    return MBaseAcc.read(AdjustedCoords);
+  }
+
+  template <typename CoordT, int Dims = Dimensions,
+            typename = detail::enable_if_t<
+                (Dims > 0) && IsValidCoordDataT<Dims, CoordT>::value>>
+  DataT read(const CoordT &Coords, const sampler &Smpl) const {
+    constexpr int AdjustedDims = Dims == 2 ? 4 : Dims + 1;
+    using CoordElemType = typename detail::TryToGetElementType<CoordT>::type;
+
+    CoordElemType ZCoord = 0;
+
+    if(std::is_same<float, CoordElemType>::value) {
+      sycl::vec<int, Dims + 1> Size = MBaseAcc.getCountInternal();
+      ZCoord = MIdx / (float)Size.template swizzle<Dims>();
+    }
+    else {
+      ZCoord = MIdx;
+    }
+
+    sycl::vec<CoordElemType, Dims> LeftoverCoords{ZCoord};
+    sycl::vec<CoordElemType, AdjustedDims> AdjustedCoords{Coords,
+                                                          LeftoverCoords};
+    return MBaseAcc.read(AdjustedCoords, Smpl);
+  }
+
+  template <typename CoordT, int Dims = Dimensions,
+            typename = detail::enable_if_t<
+                (Dims > 0) && IsValidCoordDataT<Dims, CoordT>::value>>
+  void write(const CoordT &Coords, const DataT &Color) const {
+    constexpr int AdjustedDims = Dims == 2 ? 4 : Dims + 1;
```

Comment:
this expression is used 3 times. Can you move it to separate function?

Reply:
Removed code duplication.